[28기 정해수] 메인으로 사용하는 ProductList 구현

wjdgotn77 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [x] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
1. 상품리스트에 대한 정보를 백엔드에서 받아와 리스트페이지를 만들어줍니다.
2. filter 또는 sort 버튼을 눌러 분류를 선택했을 때 , 백엔드로 해당 정보를 요청하고 필터 요소에 맞는 화면을 띄워줍니다. 
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
### 1. 상품리스트를 페이지에 띄워줍니다. 
 -  백엔드와 통신 전 목데이터를 이용해 json 파일을 fetch 받아오도록 하였습니다.
    - 넘겨받은 props를 바탕으로 상품이미지,상품명,상품가격을 표시합니다.
    - 위시리스트를 추가 기능으로 구현할 시 필요할 수 있는 아이콘을 추가하였습니다.
 -  useEffect를 이용해 해당 데이터를 state로 관리합니다.
 -  Product Component 에 props으로 해당 리스트를 넘겨주며 map을 사용해 데이터의 갯수만큼 Product Component를 그리도록 만들었습니다.
 -  Link 태그 사용하여 product-detail/아이디 값으로 해당 상품의 상세 페이지로 이동할 수 있도록 하였습니다.
 - 테스트 방법 : `/product-list` 페이지 이동

### 2. 필터 버튼 기능 구현
### ProductList.js
 1. 쿼리스트링을 사용하여 해당 url로 이동 및 GET 요청으로 데이터를 받아올 수 있는  sort 함수를 작성하였습니다.
 2. state로 관리하는 사이즈 및 가격별 값들을 useEffect의 dependency array에 넣어주어 해당 값이 바뀔 때마다 sort 함수를 호출해 화면에 렌더가 될 수 있도록 하였습니다.

### DropDown.js
 1. 사이즈 이외에 색상 혹은 카태고리 같은 다른 종류의 filter를 사용할 수 있기때문에 DropDown이라는 컴포넌트로 분리해주었습니다.
 2. 따라서 사용될 수 있는 DropDown 내의 종류에 대한 값을 STANDARD_LIST 라는 객체로 관리하고 있습니다.

### SizeFilter.js
 1. DropDown에서 받아오는 값들을 map으로 돌려 체크박스를 만들어줍니다.

- 테스트 방법 : `/product-list` 페이지 이동 후 사이즈 또는 상품가격낮은순 또는 높은순 버튼 클릭

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
[새로 알게 된 부분]
- 초기셋팅 때, 팀원들과 의견을 나누고 해당 의견들을 조합해 셋팅하는 부분이 재밌었지만 신중히 진행해야할 것 같아 어려웠습니다.
- 각 리스트를 페이지에 띄워줬을 때, 상세페이지로 어떻게 넘어가야할지에 대해 고민하였는데 , RESTful API 세션 이후 조금은 알게 된 것 같습니다.
- 마찬가지로 filter 및 sort 버튼의 기능 구현시 데이터를 받아오는 방법에 대해 알게 된 것 같습니다.
- SASS의 mixin 과 variable 을 처음 사용해보게 되어 편리함을 느꼈습니다.

[어려웠던 점]
- 쿼리스트링 ... 정말 너무 어렵습니다.. 다른 코드들을 참고하였지만, 이해하고 적용할 수 있어서 공부가 되었습니다. 다음에는 컨닝없이 로직을 스스로 구현해내고 싶습니다.
- clean all 로 초기화해주는 함수를 구현하는 부분이 어렵습니다.
- 비동기적으로 실행되는 state useEffect fetch 등이 아직도 헷갈리고 잘 모르고 사용한다고 느낍니다..

<br />

## :: 기타 질문 및 특이 사항
- 현재 맡은 부분을 모두 구현하였을 때, main 페이지를 작성해볼 것을 염두에 두고 Main에서 작업 대신 ProductList 폴더를 따로 만들게 되었습니다.
- 각 상세페이지에서 url 에 대해 접근할 수 있는 방법에 대해 찾아봐야할 것 같습니다.
- 현재 사이즈별 체크박스를 클릭했을때 filter되어 해당 사이즈만 리스트페이지에 그려주는 로직을 작성중인데, 머릿속으로 생각했던 것 만큼 로직이 간단한 것 같지 않아서 어렵게 느껴지고 처음 계획한 구현 기간보다 길어지고 있습니다.
 

✅  질문사항

🎄NEWWWWW 질문입니다.🎄

1. 현재 제가 작성중인 ProductList/component 폴더의 구조가 적절한지 궁금합니다.
  Product.js - productList 에서 map으로 뿌려주는 상품 카드 컴포넌트.
  DropDown.js - 이후 확장성을 고려 DropDown 이라고 만들어주었고 하위에 SizeFilter 라는 체크박스 관련 컴포넌트가 존재합니다.
2. DropDown, sizeFilter 의 SASS를 DropDown.scss 로 함께 관리해주고 있는데요.(각각의 내용이 많지 않아서요..) 분리해야할까요?
3. SizeFilter.js 의 label 태그의 key 값을 `( item + index )` 로 주었습니다. 고유한 값을 넣어주려다보니 이렇게 해보게 되었는데요.
 => XS01 이런식으로 되어 있는데 적절한지 궁금합니다.

---

 ~1.현재 productList 가 메인으로 뜨는 상태지만 이후 추가적으로 Main.js를 작성했을 때에는 네비게이션에서 카테고리를 선택했을 때 각 카테고리별 리스트페이지가 되어야하는데 이 부분에 대해 추가적으로 확장성을 고려한다면 현재 상황에서 생각해서 추가해야하는 부분이 있을지 궁금합니다.~

~2. FilterModal 컴포넌트 ~

```js

// FilterModal 컴포넌트

   const [selectedSize, setSelectedSize] = useState([]);
   const selectedSizeList = [...selectedSize];

   const clickSelectSize = event => {
    const { name, type, checked } = event.target;
    if (type === 'checkbox' && checked) {
      setSelectedSize([...selectedSize, name]);
    } else {
      setSelectedSize(selectedSize.filter(size => size !== name));
    }
  };
```
~위의 코드는 체크박스의 클릭여부에 따라 새로운 배열을 만들어 줍니다. UI와 관련되어지는 값이기 때문에 state 로 관리해주어야한다고 생각해서 state로 작성해주었는데요,  else 내부의 값에 filter로 클릭여부를 걸러주고 있어서 state로 관리하는 값이 맞는 것인지 궁금합니다.~

~3. productDetail에서 url의 id 값을 사용할 수 있도록 path 경로를 하단에 하나 더 작성해주었는데 , 이 방식이 맞을까요?~
```js
Router.js

function Router() {
  return (
    <BrowserRouter>
      <Nav />
      <Routes>
        <Route path="/main" element={<Main />} />
        <Route path="/login" element={<Login />} />
        <Route path="/product-list" element={<ProductList />} />
        <Route path="/product-list/:size" element={<ProductList />} /> // 이부분은 수정하겠습니다!
        <Route path="/product-detail" element={<ProductDetail />}>
          <Route path=":product_id" element={<ProductDetail />} />     // 이렇게 아이디값을 보내주는 것이 맞을까요?
        </Route>
      </Routes>
    </BrowserRouter>
  );
}
```

## 1차 리뷰 후 리뷰반영 리팩토링 내용

### Style
 1. 컴포넌트 명과 겹치는 className은 wrapper라고 추가해 적절한 네이밍으로 수정하였습니다.
 2. 일부 BEM 방식으로 작성되었던 것을 수정하였습니다.
 3. 불필요한 약어를 사용한 부분을 수정해 변수명을 변경하였습니다.

### JS
 1. 컴포넌트의 기능이 명확하도록 filterDropDown으로 컴포넌트명을 변경해주었습니다.
 2. 인라인 스타일로 작성되었던 코드를 클래스네임을 통해 display가 변경될 수 있도록 수정하였습니다.

wjdgotn77 committed:
Add: ProductList 페이지 - 상품리스트 구현

- ProductList.js
 1. 백엔드와 통신 전 목데이터를 이용해 json 파일을 fetch 받아오도록 하였습니다.
 2. useEffect를 이용해 해당 데이터를 state로 관리합니다.
 3. Product Component 에 props으로 해당 리스트를 넘겨주며 map을 사용해 데이터의 갯수만큼 Product Component를 그리도록 만들었습니다.

- Product.js
 1. 넘겨받은 props를 바탕으로 상품이미지,상품명,상품가격을 표시합니다.
 2. 위시리스트를 추가 기능으로 구현할 시 필요할 수 있는 아이콘을 추가하였습니다.

wjdgotn77 committed:
Modify: ProductList Link태그 사용하여 상세페이지 url경로 설정

-ProductList.js
 Link 태그 사용하여 product-detail/아이디 값으로 해당 상품의 상세 페이지로 이동할 수 있도록 하였습니다.

-이 외 파일들의 경우 변경사항때문에 발생한 sass 관련 수정사항입니다.

wjdgotn77 committed:
Add: ProductList 내부 Filter 버튼 및 모달창 구현중

wjdgotn77 committed:
Modify: filter버튼 및 FilteModal컴포넌트 위치 변경

- FilterModal.js
 1. 모달 내부의 체크박스의 클릭 여부에 따라 해당 사이즈만 담아주는 배열을 만들었습니다.
 2. 이 배열을 이용해 productList의 화면을 filter 해줄 수 있도록 구현 진행중입니다.
 3. 로직이 완성되면 추가적으로 중복되는 코드들을 리팩토링해줄 예정입니다.

wjdgotn77 committed:
Modify: 불필요한 코드 삭제

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+
+import './ProductList.scss';
+
+const FETCH_URL = 'http://localhost:3000/';
```
fetch url 따로 상수화해서 관리하신 것 잘 하셨습니다! 이런 데이터는 보통 config.js 등과 같은 곳에서 관리합니다. 
더불어서 해당 로컬호스트 포트번호는 계속 변할 수 있기때문에 해당 string은 따로 관리하지 않고 생략해도 좋겠네요. 
```js
const foo = () => {
	// 수정 전 fetch("http://localhost:3000/images/1")
	fetch("/images/1")
}
```

wjdgotn77 replied:
넵 신영님 테스트 할 때 사용 후에 최종에서는 지우고 올릴게요!

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+  const [isClicked, setIsClicked] = useState(false);
+
+  useEffect(() => {
+    fetch(`${FETCH_URL}/data/ProductListData/ProductListData.json`, {
+      method: 'GET',
+    })
```
fetch함수의 get method는 method의 default 값입니다. 따라서 생략이 가능합니다. 
고로 get method 요청시에는 아래와 같이 두번째 인자가 생략 가능합니다. 

```js
const foo = () => {
    fetch(`${cartAPI}/quantity`)
    .then((res) => res.json())
```

wjdgotn77 replied:
넵 ! 반영하겠습니다!

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+
+function ProductList() {
+  const [productList, setProductList] = useState([]);
+  const [isClicked, setIsClicked] = useState(false);
```
필요한 데이터들로 state 잘 만들어주셨네요! :) 좋습니다! 
다만, isClicked 라는 state명만 정확하게 작성해주시면 더 좋을 것 같습니다. 무엇을 클릭한 값인지 알 수 있도록 state명에 명시 해주세요!

wjdgotn77 replied:
네 드롭다운이 클릭된 것을 알 수 있도록 `isDropDownClicked` `setIsDropDownClicked` 로 바꿔주었어요!

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <div className="productListContent">
+      <div className="productListHeader">New Arrivals</div>
```
반복되는 'productList'라는 말은 생략해도 되겠네요!

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+        </div>
+        <FilterModal isClicked={isClicked} />
+      </div>
+      <div className="productList">
```
데이터 명과 className이 같네요. 이런 경우에는 'productList'라는 네이밍보다는 'wrapper', 'listContainer' 등과 같은 말이 더 적절할 것 같습니다.

wjdgotn77 replied:
네 신영님 `productList Wrapper` 로 수정했습니다

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,59 @@
+            <Link
+              className="productLink"
+              key={product_id}
+              to={`product-detail/${product_id}`}
```
템플릿 리터럴 잘 사용하셨네요! :) 라우터에 동적라우팅 연결도 해주셨을까요?!

wjdgotn77 replied:
아니요! 동적라우팅은 아직이에요 ㅠㅠ 해당 부분에 대해 아예 몰라서 살펴보겠습니다!

bonbon0202 commented on src/pages/ProductList/ProductList.scss:
```diff
@@ -0,0 +1,61 @@
+
+$font-main: 24px;
+$font-des: 13px;
+$font-price: 11px;
```
해당 파일에서만 사용되는 폰트 사이즈 일까요?
보통 이런 폰트 사이즈는 프로젝트 전체에 걸쳐서 이용될 확률이 높기 떄문에 팀원들과 소통 후 font관련 style 파일을 생성해서 관리하거나 variables에서 관리해도 괜찮을 것 같습니다.

wjdgotn77 replied:
넵 고려해보겠습니다 !

bonbon0202 commented on src/pages/ProductList/ProductList.scss:
```diff
@@ -0,0 +1,61 @@
+    height: 30px;
+    margin: 10px 0px;
+    padding: 10px;
+    font-size: $font-des;
```
css 작성순서도 깔끔하네요! :)

wjdgotn77 replied:
감사합니당

bonbon0202 commented on src/pages/ProductList/components/FilterModal.js:
```diff
@@ -0,0 +1,42 @@
+  const selectedSizeList = [...selectedSize];
+
+  const clickSelectSize = event => {
+    const { name, type, checked } = event.target;
```
구조분해할당도 깔끔하게 잘 해주셨네요-! :)

bonbon0202 commented on src/pages/ProductList/components/FilterModal.js:
```diff
@@ -0,0 +1,42 @@
+
+export default function FilterModal({ isClicked }) {
+  const [selectedSize, setSelectedSize] = useState([]);
+  const selectedSizeList = [...selectedSize];
```
해당 상수는 어떤 이유로 작성하셨을까요?!

wjdgotn77 replied:
엇 신영님 일단 이부분 작성한 이유는 state 값은 변경되면 안되니까 담아서 사용해야겠다고 생각을 했는데요, 아직 filter 로직을 짜는데에 왔다갔다하며 헷갈리는 부분이 많아서 조금 더 생각해보겠습니다.

bonbon0202 commented on src/pages/ProductList/components/FilterModal.js:
```diff
@@ -0,0 +1,42 @@
+    } else {
+      setSelectedSize(selectedSize.filter(size => size !== name));
+    }
+  };
```
spread operator도 잘 사용해주시고 filtering도 잘 해주셨네요! 
네 해수님. 질문 주신 내용 말씀드리면, state값으로 잘 관리해주셨습니다. 
추후 해당 값에 따라 백엔드와 통신하여 사용자에게 해당되는 데이터를 보여줄 것이기 때문에(쿼리스트링을 이용할 것입니다!) 변하는 값 state로 관리하여 백엔드와 통신할 때 사용해야합니다 :)

wjdgotn77 replied:
네 신영님 답변 감사드립니당😇

bonbon0202 commented on src/pages/ProductList/components/FilterModal.js:
```diff
@@ -0,0 +1,42 @@
+  return (
+    <form
+      className="filterModal"
+      style={isClicked ? { display: 'flex' } : { display: 'none' }}
```
- 인라인 스타일 사용은 반드시 필요한 경우가 아니면 지양해주세요
1. 우선순위가 !important 제외하고 가장 높게 적용됨
2. css관련 로직이 css 파일이 아닌 html에 있어서 유지보수하기 좋지않음
- 위 이유들로 인해서 반드시 필요한 경우가 아닌 한 인라인 스타일은 지양해주세요

위 경우, 조건에 따라 다른 Ui를 보여주고 싶다면 인라인 스타일이 아니라, 어떤 속성을 이용하면 좋을까요?
더불어 위와 같은 경우 아래와 같이 작성해줄 수도 있습니다.
```suggestion
      style={{ display: isClicked ? 'flex' : 'none' }}
```

wjdgotn77 replied:
네 신영님 인라인스타일에 대한 내용 확인했습니다.
```js
<form
      className={`filterDropDownWrapper ${isClicked ? 'show' : 'hide'}`}
```
sass 에서 show 와 hide 에 대한 display 값을 미리주고 클래스네임으로 핸들링하는 방법을 사용하는 방식으로 리팩토링 해봤어요!

bonbon0202 commented on src/pages/ProductList/components/FilterModal.js:
```diff
@@ -0,0 +1,42 @@
+import React, { useState } from 'react';
+import './FilterModal.scss';
+
+export default function FilterModal({ isClicked }) {
```
해당 ui는 모달보다는 드롭다운에 가깝습니다. 
모달창은 기존의 브라우저 페이지 위에 새로운 레이어를 까는 것을 말합니다. 그래서 다음 진행으로 넘어가기 위한 필요에 의해 사용되는 창이라고 생각하시면 됩니다. 전체 페이지를 막아서 해당 모달창에 대한 내용을 진행해야 다음 페이지로 넘어갈 수 있는 형태입니다. 해당 ui는 모달창보다는 드롭다운에 가깝네요!

wjdgotn77 replied:
네 신영님 수정했습니다👌

bonbon0202 commented on src/pages/ProductList/components/FilterModal.scss:
```diff
@@ -0,0 +1,25 @@
+@import '../../../styles/mixin';
+@import '../../../styles/variables';
+
+$font-des: 13px;
```
불필요한 약어보다는 description으로 표현해도 괜찮을것 같습니다.

wjdgotn77 replied:
넵

bonbon0202 commented on src/pages/ProductList/components/FilterModal.scss:
```diff
@@ -0,0 +1,25 @@
+.filterModal {
+  position: absolute;
+  flex-direction: column;
+  top: calc($font-des + 20px);
```
폰트사이즈 크기를 기준으로 값을 더해주신 이유가 있으실까요?!

wjdgotn77 replied:
네네 해당 드롭다운이 상위 button 의 폰트 크기보다 20px 떨어진 곳에 그려지길 원해서 이렇게 작성했는데요, 더 나은 방법이 있을까요?

bonbon0202 commented on src/pages/ProductList/components/Product.js:
```diff
@@ -0,0 +1,24 @@
+  return (
+    <div>
+      <div className="imageHolder">
+        <img alt={`${name}`} src={url} />
```
특별히 string과 JS가 합쳐져 사용되지 않기 때문에 아래와 같이 작성해도 되겠네요!
```suggestion
        <img alt={name} src={url} />
```

bonbon0202 commented on src/pages/ProductList/components/Product.js:
```diff
@@ -0,0 +1,24 @@
+        <img alt={`${name}`} src={url} />
+      </div>
+      <div className="description">
+        <div className="description__name">{name}</div>
+        <div className="description__price">{price} 원</div>
+        <div className="description__color">black</div>
+        <div className="description__wishButton">
```
해당 부분에서는 네이밍 형식이 달라진 이유가 있으실까요?!

wjdgotn77 replied:
앗 신영님 해당 description 의 종류들이라고 생각해서 저렇게 작성했는데, 지금 다시 보니 nesting해서 적어준다면 그냥 name, price, color, wishButton 으로 클래스네임을 바꿔줘도 되겠군요! 수정하였습니당

bonbon0202 reviewed (commented):
해수님! 코드를 깔끔하게 잘 작성하시는 편이시네요! :) 
리뷰 달아드린 내용 참조해서 리팩토링 하시기를 바랍니다. 
더불어서 추후에 메인 페이지가 추가가 된다면 그때 카테고리에 리스트 페이지로 연결하면 됩니다. 그 작업은 추후에 하셔도 괜찮을 것 같네요!
그 외 다른 질문은 코멘트로 달아드렸는데, 이해가 잘 안되신다면 금일 줌 QnA에서 질문 진행해주세요!

wjdgotn77 committed:
Refactor: PR 코드리뷰 반영 리팩토링

- Style
 1. 컴포넌트 명과 겹치는 className은 wrapper라고 추가해 적절한 네이밍으로 수정하였습니다.
 2. 일부 BEM 방식으로 작성되었던 것을 수정하였습니다.
 3. 불필요한 약어를 사용한 부분을 수정해 변수명을 변경하였습니다.

- JS
 1. 컴포넌트의 기능이 명확하도록 filterDropDown으로 컴포넌트명을 변경해주었습니다.
 2. 인라인 스타일로 작성되었던 코드를 클래스네임을 통해 display가 변경될 수 있도록 수정하였습니다.

wjdgotn77 committed:
Add: Router.js에 productDetail페이지 추가

wjdgotn77 committed:
Modify: 드롭다운필터기능 구현중

wjdgotn77 committed:
Modify: useLocation 위치변경 - 드롭다운필터기능 구현중

wjdgotn77 committed:
Modify: 드롭다운필터기능 구현중

wjdgotn77 committed:
Modify: 사이즈 필터 구현 완료, 가격순 구현 중

bonbon0202 commented on src/Router.js:
```diff
@@ -12,7 +13,11 @@ function Router() {
         <Route path="/login" element={<Login />} />
-        <Route path="/product-detail" element={<ProductDetail />} />
+        <Route path="/product-list" element={<ProductList />} />
+        <Route path="/product-list/:size" element={<ProductList />} />
```
size는 화면에 display 되는 grid 개수를 의미하는 걸까요?

wjdgotn77 replied:
수정하였습니다!

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,89 @@
+
+  const navigate = useNavigate();
+
+  function sort() {
```
아래에서는 arrow function 사용해주고 계셔서 통일해주시면 좋을 것 같아요! 컴포넌트 내부 함수들은 arrow function으로 해도 괜찮겠네요!

wjdgotn77 replied:
넵 통일하겠습니다~!

bonbon0202 commented on src/pages/ProductList/components/DropDown.js:
```diff
@@ -0,0 +1,35 @@
+  const [isDropDownActive, setIsDropDownActive] = useState(false);
+
+  return (
+    <div>
```
최상단 요소의 className으로 네스팅하는 것 있지 말아주세요! 나중에 다른 팀원들 코드랑 merge 될때 css 충돌이 생길 수 있습니다.

wjdgotn77 replied:
네! dropDown으로 className 추가해주었습니다!

bonbon0202 commented on src/pages/ProductList/components/DropDownBox.js:
```diff
@@ -0,0 +1,27 @@
+
+  return (
+    <div className={`${dropDownStyle} ${showDropDown}`}>
+      {standardList.map((item, idx) => (
+        <label key={idx} onChange={checkedSize}>
+          <input type="checkbox" standard={standard} value={item} />
+          {item}
+        </label>
+      ))}
```
이제 map 메서드 깔끔하게 잘 작성해주시네요! :)

bonbon0202 reviewed (commented):
오 :) 해수님 코드가 많이 깔끔해지신것 같아요!!
쿼리스트링 백엔드 분과도 한번 맞추어 보시고, (맞추시면서 조금 수정해야할 부분이 있으실거예요) 
추가 구현 진행 계속 해주세요!

wjdgotn77 committed:
Modify: 사이즈 및 가격별 필터 구현 완료.

-ProductList.js
 1. 쿼리스트링을 사용하여 해당 url로 이동 및 GET 요청으로 데이터를 받아올 수 있는  sort 함수를 작성하였습니다.
 2. state로 관리하는 사이즈 및 가격별 값들을 useEffect의 dependency array에 넣어주어 해당 값이 바뀔 때마다 sort 함수를 호출해 화면에 렌더가 될 수 있도록 하였습니다.

-DropDown.js
 1. 사이즈 이외에 색상 혹은 카태고리 같은 다른 종류의 filter를 사용할 수 있기때문에 DropDown이라는 컴포넌트로 분리해주었습니다.
 2. 따라서 사용될 수 있는 DropDown 내의 종류에 대한 값을 STANDARD_LIST 라는 객체로 관리하고 있습니다.

-SizeFilter.js
 1. DropDown에서 받아오는 값들을 map으로 돌려 체크박스를 만들어줍니다.

wjdgotn77 committed:
Modify: 체크박스 내부 사이즈 key값 고유하게 변경

wjdgotn77 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,114 @@
+
+  useEffect(() => {
+    getDataBySort();
+  }, [selectSizeList, selectSort]);
```
<img width="677" alt="스크린샷 2022-01-04 오전 12 45 48" src="https://user-images.githubusercontent.com/80671448/147950701-958e1a17-d408-4e2d-a2db-5e24231167e9.png">

@bonbon0202  신영님 이부분 의존성 배열에 getDateBySort를 넣어주거나 useCallBack을 사용해줘야하는 것 같은데, 제가 생각한 것이 맞을까요? 그렇다면 두 방법중 의존성 배열에 해당 함수를 넣어주는 것으로 처리해도 괜찮을까요?
=> 엇,,,이렇게 했다가 무한렌더링의 늪에 빠졌습니다 ...😂

만약 위의 생각이 틀렸다면, https://github.com/facebook/react/issues/14920 이곳을 참고해보면 되는걸까요?

bonbon0202 replied:
이 부분 내일 라이브 코드리뷰 때 같이 보도록 하겠습니다!

wjdgotn77 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,114 @@
+    const queryString = filterQuery.concat(sortQuery);
+    navigate(`?${queryString}`);
+
+    fetch(`?${queryString}`, {
+      headers: {
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+      },
+    })
+      .then(res => res.json())
+      .then(data => {
+        setProductList(data);
+      });
+  };
```
<img width="708" alt="스크린샷 2022-01-04 오전 12 59 57" src="https://user-images.githubusercontent.com/80671448/147952161-2fc1eb20-923e-4350-ab5e-33b25105c411.png">

@bonbon0202  신영님 이 부분 fetch 주소를 보니 제가 작성한 코드로 인해 `?`가 붙어나오는데.. ㅜ ㅜ 이부분 도대체 어떻게 해야 .. `?`를 핸들링 할 수 있을까요 ...

bonbon0202 replied:
어떻게 해야할까요..?ㅎㅎㅎ
지금 fetch 주소에 `?${queryString}` 으로 무조건 '?'이 포함되게 되어있네요. 조건에 따라 쿼리스트링이 포함될때에만 '?'가 같이 붙어서 들어가야 하지 않을까요?! 
아니면 페이지가 새로고침 되었을 때, query parameter를 가져오는 훅을 이용해서 붙인 다음에 fetch를 해도 되겠네요!

wjdgotn77 committed:
Add: filter 리셋 기능 버튼 구현 완료

wjdgotn77 committed:
Modify: sort 쿼리스트링 앤드심볼 조건 추가

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,141 @@
+    if (sizeType && checked) {
+      setSelectSizeList(size => [...size, value]);
+    } else {
+      setSelectSizeList(size => size.filter(el => el !== value));
```
.

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,141 @@
+      } else {
+        sortQuery = `order=price_${selectSort}`;
+      }
+    }
```
.

bonbon0202 commented on src/pages/ProductList/ProductList.js:
```diff
@@ -0,0 +1,141 @@
+    setSelectSort(className);
+  };
+
+  const checkedSize = event => {
```
.

wjdgotn77 committed:
Modify: 백 통신 후 에러 핸들링을 위한 코드 수정

wjdgotn77 committed:
Add: 무한스크롤 구현 - 백과 통신완료

wjdgotn77 committed:
Modify: clear버튼 클릭 시 checkbox 안 닫히는 부분 수정 완료

wjdgotn77 committed:
Modify: 무한스크롤 추가 수정 및 로딩페이지 추가
